### DISCUSSION TITLE
Lab 02.03 Task 3 Fill in missing data

### DISCUSSION THREAD

**Discussion by jhondbv** [rramosp/ai4eng.v1#226] [02 - PYTHON]
Hola , para la task3 tengo el codigo funcionando correctamente cuando lo ejecuto manualmente , pero cuando le doy submit me genera un error , este es el codigo : 
```python
def fillna(df):
    # make sure to make a copy in case you modify the original df
    d = df.copy()
    # tuve que agregar una validacion para que me eliminara duplicados , aun no entiendo
    # por que me generaba un error de indices duplicados 
    d = d.loc[~d.index.duplicated(keep='first')] 
    price_mean = d['price'].mean() #promedio de datos
    price_std = d['price'].std() # deviacion estandar , para saber cuanto varian los datos
    missing= d['price'].isna().sum() # numero de datos NaN
    samples = np.random.normal(price_mean,price_std,missing) # Generando datos aleatorios 
    new_price = pd.Series(samples,index=d.index[d['price'].isna()])# pd series para los indices donde estan los datos NaN
    d['price'].fillna(new_price,inplace=True)
    d['price']= d['price'].round(2)
    return d
```

cuando lo corro manualmente con el siguiente comando , me retorna el dataframe con las mismas columnas que fue creado , el mismo tipo de datos y los mismos decimales 
```python
d = create_df(missing=True)
fillna(d)
```

pero cuando le doy submmit al codigo , me genera el siguiente error :
testing with 100 random calls
dataframe shape is incorrect

![image](https://user-images.githubusercontent.com/12846305/220017220-1986461a-71e2-4fe2-a60b-77a0dde41cd1.png)
![image](https://user-images.githubusercontent.com/12846305/220017262-8ec2f495-ed61-41fe-8d64-067137489c07.png)

**Comment by jhondbv** [comment]
le hice una pequeña modificacion al codigo con una sugerencia hecha por el profesor en otro post (https://github.com/rramosp/ai4eng.v1/discussions/15#discussioncomment-795820)
asi quedo el codigo pero aun sigue sin funcionar cuando lo envio a calificar 
```python
def fillna(df):
    # make sure to make a copy in case you modify the original df
    d = df.copy()
    d = d.loc[~d.index.duplicated(keep='first')]
    price_mean = d['price'].mean() #promedio de datos
    price_std = d['price'].std() # deviacion estandar , para saber cuanto varian los datos
    missing= d['price'].isna().sum() # numero de datos NaN
    samples = np.random.normal(price_mean,price_std,missing) # Generando datos aleatorios 
    new_price = pd.Series(samples,index=d.index[d['price'].isna()])# pd series para los indices donde estan los datos NaN
    # d['price'].fillna(new_price,inplace=True)
    d['price'][d.price.isna()] = new_price
    d['price']= d['price'].round(2)
    return d
```

**Reply by Felipe-RA** [reply]
Saludos @jhondbv 

```python
def fillna(df):
    # make sure to make a copy in case you modify the original df
    d = df.copy()
   --------> d = d.loc[~d.index.duplicated(keep='first')]
    price_mean = d['price'].mean() #promedio de datos
    price_std = d['price'].std() # deviacion estandar , para saber cuanto varian los datos
    missing= d['price'].isna().sum() # numero de datos NaN
    samples = np.random.normal(price_mean,price_std,missing) # Generando datos aleatorios 
    new_price = pd.Series(samples,index=d.index[d['price'].isna()])# pd series para los indices donde estan los datos NaN
    # d['price'].fillna(new_price,inplace=True)
    d['price'][d.price.isna()] = new_price
    d['price']= d['price'].round(2)
    return d
```

La línea tres de tu función (que acabo de marcar con una flecha en el código de arriba) es la fuente del problema.

Cuando copias un `dataframe` con `copy()` también copias su índice.

Esta línea causa el problema pues, aunque el contenido del dataframe es correcto, el índice no conforma con las dimensiones iniciales (aunque a simple vista parezca que lo hace, echale un vistazo a lo que devuelve la función [.Index.duplicated](https://pandas.pydata.org/docs/reference/api/pandas.Index.duplicated.html))

Cuentame si con esto logras solucionar el problema, estoy pendiente

¡Hasta pronto!

**Reply by jhondbv** [reply]
Hola Felipe , si , eso soluciono el error , Gracias
